### PR TITLE
Update openblas to build with apple-clang@12.0.0

### LINKED ISSUE #18658: Installation issue: openblas build failed
### Steps to reproduce the issue

```console
$ spack install -v openblas
```
Error messages are shown [here](https://pastebin.com/raw/qeaiaPHe).

### Information on your system

* **Spack:** 0.15.4-947-1aceb38b8
* **Python:** 3.8.5
* **Platform:** darwin-macos-skylake
* **Version:** 11.0 beta Big Sur

* `packages.yaml` [here](https://pastebin.com/raw/qt26tVqZ)

### Additional information

* [spack-build-out.txt](https://github.com/spack/spack/files/5214484/spack-build-out.txt)
* [spack-build-env.txt](https://github.com/spack/spack/files/5214482/spack-build-env.txt)

### General information

- [x] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [x] I have run `spack maintainers <name-of-the-package>` and @mentioned any maintainers
- [x] I have uploaded the build log and environment files
- [x] I have searched the issues of this repo and believe this is not a duplicate

### LINKED ISSUE #19041: Installation issue: openblas@0.3.10 on MacOSX
After having handed in my thesis (and therefore can risk screwing up my environment again) I tried to upgrade to the up-to-date spack/develop branch. Turns out this breaks building `openblas`. I'm not sure why, there are at least 3 option that I'm aware of:
1) The latest Xcode update (to `apple-clang@12.0.0`) is unable to build `openblas` (previously it was `clang@11.0.0-apple`, i.e. my software is older than the clang-rename)
2) Neither the old, nor the new `gfortran` binary is working with the new system clang (I tried my old one from a `gcc@9.3.0%clang@11.0.0-apple` as well as a new one from `gcc@10.3.0%apple-clang@12.0.0`)
3) Something broke in the `openblas/package.py` during the update

I'm using the same environment as before with a couple of not-in-upstream-spack-packages but non of those can interfere (mostly python packages far above blas). Unfortunately I cannot easily recover the old base commit, but the current one is `04771ad9f8` from just now.

Currently I'm trying to set `{blas,lapack}: [libvecfort]` as suggested in #7803 and it seems to work (at least `py-numpy` successfully installed). I'm attaching the build log and environment in the hope that someone can point out where I'm just being stupid and that this build normally works... Googling the error message didn't turn up any information, since this (see below) is "just" an `Implicit declaration of function 'cggsvp_' is invalid in C99` it may be fine to just patch out the `-Werror` flag and be done with it but I (at least currently) don't want to fiddle with the build and hope for the `libvecfort` thing to work.

Pinging a couple of people that work on MacOSX for their comments @adamjstewart @davydden ? 

```
% spack debug report
* **Spack:** 0.15.4-1203-1240db659a
* **Python:** 3.7.4
* **Platform:** darwin-catalina-skylake
```

[spack-build-out.txt](https://github.com/spack/spack/files/5296990/spack-build-out.txt)
[spack-build-env.txt](https://github.com/spack/spack/files/5296994/spack-build-env.txt)

```
healther@MacBook-Air-2 spack % spack install
==> Installing environment generalenv
==> Installing openblas
==> No binary for openblas found: installing from source
==> Warning: microarchitecture specific optimizations are not supported yet on mixed compiler toolchains [check apple-clang@12.0.0 for further details]
==> Using cached archive: /Users/healther/Documents/git.nosync/spack/var/spack/cache/_source-cache/archive/04/0484d275f87e9b8641ff2eecaa9df2830cbe276ac79ad80494822721de6e1693.tar.gz
==> openblas: Executing phase: 'edit'
==> openblas: Executing phase: 'build'
==> Error: ProcessError: Command exited with status 2:
    'make' '-j4' 'CC=/Users/healther/Documents/git.nosync/spack/lib/spack/env/clang/clang' 'FC=/Users/healther/Documents/git.nosync/spack/lib/spack/env/clang/gfortran' 'MAKE_NB_JOBS=0' 'ARCH=x86_64' 'DYNAMIC_ARCH=1' 'USE_OPENMP=0' 'USE_THREAD=0' 'libs' 'netlib' 'shared'

9 errors found in build log:
     19707    /Users/healther/Documents/git.nosync/spack/lib/spack/env/clang/clang -O2 -DMAX_STACK_ALLOC=2048 -DEXPRECISION -Wall -m64 -DF_INTERFACE_GFORT -fPIC -DDYNAMIC_ARCH -DNO_WARMUP -DMAX_CPU_NUMBER=4 -DMAX_PARALLEL_NUMBER=1 -DVERSION="0.3.10" -UASMNAME -UASMFNAME 
              -UNAME -UCNAME -UCHAR_NAME -UCHAR_CNAME -DASMNAME=_ -DASMFNAME=__ -DNAME=_ -DCNAME= -DCHAR_NAME="_" -DCHAR_CNAME="" -DNO_AFFINITY -I. -DHAVE_LAPACK_CONFIG_H -I../include -c -o lapacke_zupmtr_work.o lapacke_zupmtr_work.c
     19708    /Users/healther/Documents/git.nosync/spack/lib/spack/env/clang/clang -O2 -DMAX_STACK_ALLOC=2048 -DEXPRECISION -Wall -m64 -DF_INTERFACE_GFORT -fPIC -DDYNAMIC_ARCH -DNO_WARMUP -DMAX_CPU_NUMBER=4 -DMAX_PARALLEL_NUMBER=1 -DVERSION="0.3.10" -UASMNAME -UASMFNAME 
              -UNAME -UCNAME -UCHAR_NAME -UCHAR_CNAME -DASMNAME=_ -DASMFNAME=__ -DNAME=_ -DCNAME= -DCHAR_NAME="_" -DCHAR_CNAME="" -DNO_AFFINITY -I. -DHAVE_LAPACK_CONFIG_H -I../include -c -o lapacke_cggsvp.o lapacke_cggsvp.c
     19709    /Users/healther/Documents/git.nosync/spack/lib/spack/env/clang/clang -O2 -DMAX_STACK_ALLOC=2048 -DEXPRECISION -Wall -m64 -DF_INTERFACE_GFORT -fPIC -DDYNAMIC_ARCH -DNO_WARMUP -DMAX_CPU_NUMBER=4 -DMAX_PARALLEL_NUMBER=1 -DVERSION="0.3.10" -UASMNAME -UASMFNAME 
              -UNAME -UCNAME -UCHAR_NAME -UCHAR_CNAME -DASMNAME=_ -DASMFNAME=__ -DNAME=_ -DCNAME= -DCHAR_NAME="_" -DCHAR_CNAME="" -DNO_AFFINITY -I. -DHAVE_LAPACK_CONFIG_H -I../include -c -o lapacke_cggsvp_work.o lapacke_cggsvp_work.c
     19710    /Users/healther/Documents/git.nosync/spack/lib/spack/env/clang/clang -O2 -DMAX_STACK_ALLOC=2048 -DEXPRECISION -Wall -m64 -DF_INTERFACE_GFORT -fPIC -DDYNAMIC_ARCH -DNO_WARMUP -DMAX_CPU_NUMBER=4 -DMAX_PARALLEL_NUMBER=1 -DVERSION="0.3.10" -UASMNAME -UASMFNAME 
              -UNAME -UCNAME -UCHAR_NAME -UCHAR_CNAME -DASMNAME=_ -DASMFNAME=__ -DNAME=_ -DCNAME= -DCHAR_NAME="_" -DCHAR_CNAME="" -DNO_AFFINITY -I. -DHAVE_LAPACK_CONFIG_H -I../include -c -o lapacke_dggsvp.o lapacke_dggsvp.c
     19711    /Users/healther/Documents/git.nosync/spack/lib/spack/env/clang/clang -O2 -DMAX_STACK_ALLOC=2048 -DEXPRECISION -Wall -m64 -DF_INTERFACE_GFORT -fPIC -DDYNAMIC_ARCH -DNO_WARMUP -DMAX_CPU_NUMBER=4 -DMAX_PARALLEL_NUMBER=1 -DVERSION="0.3.10" -UASMNAME -UASMFNAME 
              -UNAME -UCNAME -UCHAR_NAME -UCHAR_CNAME -DASMNAME=_ -DASMFNAME=__ -DNAME=_ -DCNAME= -DCHAR_NAME="_" -DCHAR_CNAME="" -DNO_AFFINITY -I. -DHAVE_LAPACK_CONFIG_H -I../include -c -o lapacke_dggsvp_work.o lapacke_dggsvp_work.c
     19712    /Users/healther/Documents/git.nosync/spack/lib/spack/env/clang/clang -O2 -DMAX_STACK_ALLOC=2048 -DEXPRECISION -Wall -m64 -DF_INTERFACE_GFORT -fPIC -DDYNAMIC_ARCH -DNO_WARMUP -DMAX_CPU_NUMBER=4 -DMAX_PARALLEL_NUMBER=1 -DVERSION="0.3.10" -UASMNAME -UASMFNAME 
              -UNAME -UCNAME -UCHAR_NAME -UCHAR_CNAME -DASMNAME=_ -DASMFNAME=__ -DNAME=_ -DCNAME= -DCHAR_NAME="_" -DCHAR_CNAME="" -DNO_AFFINITY -I. -DHAVE_LAPACK_CONFIG_H -I../include -c -o lapacke_sggsvp.o lapacke_sggsvp.c
  >> 19713    lapacke_cggsvp_work.c:52:9: error: implicit declaration of function 'cggsvp_' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
     19714            LAPACK_cggsvp( &jobu, &jobv, &jobq, &m, &p, &n, a, &lda, b, &ldb, &tola,
     19715            ^
     19716    ../include/lapack.h:3771:23: note: expanded from macro 'LAPACK_cggsvp'
     19717    #define LAPACK_cggsvp LAPACK_GLOBAL(cggsvp,CGGSVP)
     19718                          ^
     19719    ../include/lapacke_mangling.h:12:39: note: expanded from macro 'LAPACK_GLOBAL'

     ...

     19741    ../include/lapacke_mangling.h:12:39: note: expanded from macro 'LAPACK_GLOBAL'
     19742    #define LAPACK_GLOBAL(lcname,UCNAME)  lcname##_
     19743                                          ^
     19744    <scratch space>:94:1: note: expanded from here
     19745    cggsvp3_
     19746    ^
  >> 19747    lapacke_cggsvp_work.c:139:9: error: implicit declaration of function 'cggsvp_' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
     19748            LAPACK_cggsvp( &jobu, &jobv, &jobq, &m, &p, &n, a_t, &lda_t, b_t,
     19749            ^
     19750    ../include/lapack.h:3771:23: note: expanded from macro 'LAPACK_cggsvp'
     19751    #define LAPACK_cggsvp LAPACK_GLOBAL(cggsvp,CGGSVP)
     19752                          ^
     19753    ../include/lapacke_mangling.h:12:39: note: expanded from macro 'LAPACK_GLOBAL'
     19754    #define LAPACK_GLOBAL(lcname,UCNAME)  lcname##_
     19755                                          ^
     19756    <scratch space>:229:1: note: expanded from here
     19757    cggsvp_
     19758    ^
     19759    2 errors generated.
  >> 19760    make[3]: *** [lapacke_cggsvp_work.o] Error 1
     19761    make[3]: *** Waiting for unfinished jobs....
  >> 19762    lapacke_dggsvp_work.c:48:9: error: implicit declaration of function 'dggsvp_' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
     19763            LAPACK_dggsvp( &jobu, &jobv, &jobq, &m, &p, &n, a, &lda, b, &ldb, &tola,
     19764            ^
     19765    ../include/lapack.h:3763:23: note: expanded from macro 'LAPACK_dggsvp'
     19766    #define LAPACK_dggsvp LAPACK_GLOBAL(dggsvp,DGGSVP)
     19767                          ^
     19768    ../include/lapacke_mangling.h:12:39: note: expanded from macro 'LAPACK_GLOBAL'

     ...

     19790    ../include/lapacke_mangling.h:12:39: note: expanded from macro 'LAPACK_GLOBAL'
     19791    #define LAPACK_GLOBAL(lcname,UCNAME)  lcname##_
     19792                                          ^
     19793    <scratch space>:95:1: note: expanded from here
     19794    dggsvp3_
     19795    ^
  >> 19796    lapacke_dggsvp_work.c:127:9: error: implicit declaration of function 'dggsvp_' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
     19797            LAPACK_dggsvp( &jobu, &jobv, &jobq, &m, &p, &n, a_t, &lda_t, b_t,
     19798            ^
     19799    ../include/lapack.h:3763:23: note: expanded from macro 'LAPACK_dggsvp'
     19800    #define LAPACK_dggsvp LAPACK_GLOBAL(dggsvp,DGGSVP)
     19801                          ^
     19802    ../include/lapacke_mangling.h:12:39: note: expanded from macro 'LAPACK_GLOBAL'
     19803    #define LAPACK_GLOBAL(lcname,UCNAME)  lcname##_
     19804                                          ^
     19805    <scratch space>:229:1: note: expanded from here
     19806    dggsvp_
     19807    ^
     19808    2 errors generated.
  >> 19809    make[3]: *** [lapacke_dggsvp_work.o] Error 1
  >> 19810    make[2]: *** [lapacke] Error 2
  >> 19811    make[1]: *** [lapackelib] Error 2
  >> 19812    make: *** [netlib] Error 2

See build log for details:
  /var/folders/1q/jxx1cgcx3l13d153s6_c1zc00000gn/T/healther/spack-stage/spack-stage-openblas-0.3.10-gkmqerz6bgkp6uw5jdkmeiiadxlwyyhw/spack-build-out.txt
```

### Steps to reproduce the issue

```console
$ spack install openblas
...
```

### Information on your system

```
spack:
  # add package specs to the `specs` list
  specs: [py-notebook ^python@3.7.4, py-numpy, py-scipy, nest~mpi~optimize~openmp,
    py-matplotlib, py-ipython, py-line-profiler, py-memory-profiler, boost, yaml-cpp,
    ccache, py-pudb, htop, py-virtualenv, py-sympy, py-scikit-optimize, py-scikit-learn,
    py-scikit-image, py-requests, py-flake8, py-black, py-pandas, py-numba, py-tornado,
    py-ics, py-xlrd, py-git-review, py-seaborn, openssh, py-pynn, py-sbs]
  view: true
  config: {}
  mirrors: {}
  modules:
    enable: []
  packages: {}
  repos: []
  upstreams: {}
  definitions: []
  concretization: together
  compilers:
  - compiler:
      spec: apple-clang@12.0.0
      paths:
        cc: /usr/bin/clang
        cxx: /usr/bin/clang++
        f77: /Users/healther/Documents/git.nosync/spack/opt/spack/darwin-catalina-x86_64/apple-clang-12.0.0/gcc-10.2.0-j73olo3b5grspw4skyyfkynhtnagj3hd/bin/gfortran
        fc: /Users/healther/Documents/git.nosync/spack/opt/spack/darwin-catalina-x86_64/apple-clang-12.0.0/gcc-10.2.0-j73olo3b5grspw4skyyfkynhtnagj3hd/bin/gfortran
      flags: {}
      operating_system: catalina
      target: x86_64
      modules: []
      environment: {unset: []}
      extra_rpaths: []
```

### PR DESCRIPTION
**Pull request by healther**
fixes #19041 
fixes #18658

tested on 
 % spack debug report
* **Spack:** 0.15.4-1204-15c41ed877
* **Python:** 3.7.4
* **Platform:** darwin-catalina-skylake

See discussion in https://github.com/xianyi/OpenBLAS/issues/2870 and comments in the diff

### CODE REVIEW

File: var/spack/repos/builtin/packages/openblas/package.py [hunk 1]
```diff
@@ -96,6 +96,12 @@ class Openblas(MakefilePackage):
     # Fix ICE in LLVM 9.0.0 https://github.com/xianyi/OpenBLAS/pull/2329
     # Patch as in https://github.com/xianyi/OpenBLAS/pull/2597
     patch('openblas_appleclang11.patch', when='@0.3.8:0.3.9 %apple-clang@11.0.3')
+    # There was an error in Reference-LAPACK that is triggeret by Xcode12
+    # fixed upstream by https://github.com/xianyi/OpenBLAS/pull/2808 and
+    # should be included in post 0.3.10 versions. Application to earlier
+    # versions was not tested.
+    # See also https://github.com/xianyi/OpenBLAS/issues/2870
+    patch('openblas-0.3.10-appleclang12.patch', when='@:0.3.10 %apple-clang@12.0.0:')
```
Comment (dennisklein) [inline var/spack/repos/builtin/packages/openblas/package.py]:
```suggestion
    patch('openblas-0.3.10-appleclang12.patch', when='@0.3.8:0.3.10 %apple-clang@12:')
```

Just tried with `spack patch openblas@...` and the patch is only applicable @ `0.3.[8-10]`.

File: var/spack/repos/builtin/packages/openblas/openblas-0.3.10-appleclang12.patch [hunk 2]
```diff
@@ -0,0 +1,384 @@
+From 1a05523709567d54273fe1000ae4319df437737c Mon Sep 17 00:00:00 2001
```
Comment (dennisklein) [inline var/spack/repos/builtin/packages/openblas/openblas-0.3.10-appleclang12.patch]: Please generate the patch file from this commit: https://github.com/xianyi/OpenBLAS/commit/f42e84d46c52f4ee1e05af8f365cd85de8a77b95

The one you used appears to have been a temporary one which does not belong to the `develop` branch of OpenBLAS.

Comment (adamjstewart) [reply var/spack/repos/builtin/packages/openblas/openblas-0.3.10-appleclang12.patch]: You can also add the patch via URL and sha256.

File: var/spack/repos/builtin/packages/openblas/package.py (hunk 1 above)
Comment (healther) [reply var/spack/repos/builtin/packages/openblas/package.py]: done

File: var/spack/repos/builtin/packages/openblas/openblas-0.3.10-appleclang12.patch (hunk 2 above)
Comment (healther) [reply var/spack/repos/builtin/packages/openblas/openblas-0.3.10-appleclang12.patch]: I only tested that the patch applies and did not do another test build, I assume it still works fine, but can do another test build tomorrow at the earliest. Do you want me to or merge as is?

Comment (dennisklein) [reply var/spack/repos/builtin/packages/openblas/openblas-0.3.10-appleclang12.patch]: > You can also add the patch via URL and sha256.

Just for my understanding: Will a URL-based patch be included in the source cache?

Comment (adamjstewart) [reply var/spack/repos/builtin/packages/openblas/openblas-0.3.10-appleclang12.patch]: Yes, a URL-based patch would be cached just like resources.

**Review by dennisklein — APPROVED** [review]

**Comment by healther** [comment]
Test build complete and working

**Review by scheibelp — APPROVED** [review]

**Pull request merged**